crates/subspace-solving/src/codec.rs

```diff
@@ -16,133 +16,183 @@
-    //
-    // // TODO: Refactor from being CUDA-specific to be batch-oriented
-    // /// Encode given batch of pieces using GPU, and CPU for the leftovers
-    // #[cfg(feature = "cuda")]
-    // pub fn cuda_batch_encode(&self, pieces: &mut [u8], nonce_array: &[u64]) {
-    //     // each expanded_iv will be in format [u8; 32], so `piece_amount` expanded_iv's
-    //     // should consume [u8; 32 * piece_amount] space.
-    //     let piece_count = pieces.len() / PIECE_SIZE;
-    //     let mut expanded_iv_vector: Vec<u8> = Vec::with_capacity(piece_count * PRIME_SIZE);
-    //     let mut expanded_iv;
-    //     for nonce in nonce_array {
-    //         // same farmer_public_key_hash will be used for each expanded_iv
-    //         expanded_iv = self.farmer_public_key_hash;
-    //
-    //         // select the nonce from nonce_array, xor it with the farmer_public_key_hash
-    //         nonce
-    //             .to_le_bytes()
-    //             .iter()
-    //             .rev()
-    //             .zip(expanded_iv.iter_mut().rev())
-    //             .for_each(|(nonce_byte, expanded_iv_byte)| *expanded_iv_byte ^= nonce_byte);
-    //         //*/
-    //         expanded_iv_vector.extend(expanded_iv);
-    //     }
-    //
-    //     // If there any leftovers from 1024x pieces, cpu will handle them
-    //     let cpu_encode_end_index = piece_count % GPU_PIECE_BLOCK;
-    //
-    //     // CPU encoding:
-    //     for x in 0..cpu_encode_end_index {
-    //         cpu::encode(
-    //             &mut pieces[x * PIECE_SIZE..(x + 1) * PIECE_SIZE],
-    //             &expanded_iv_vector[x * PRIME_SIZE..(x + 1) * PRIME_SIZE],
-    //             ENCODE_ROUNDS,
-    //         )
-    //         .unwrap();
-    //     }
-    //
-    //     // GPU encoding:
-    //     cuda::encode(
-    //         &mut pieces[cpu_encode_end_index * PIECE_SIZE..],
-    //         &expanded_iv_vector[cpu_encode_end_index * PRIME_SIZE..],
-    //         ENCODE_ROUNDS,
-    //     )
-    //     .unwrap();
-    // }
+    /// Encode given batch of pieces using the best method available, which might be GPU, CPU or
+    /// combination of both.
+    ///
+    /// [`SubspaceCodec::recommended_batch_size()`] can be used to determine the recommended batch
+    /// size, input should ideally contain at least that many worth of pieces to achieve highest
+    /// efficiency, it is recommended that the input is a multiple of that, but, strictly speaking,
+    /// doesn't have to be.
+    #[allow(unused_mut)]
+    pub fn batch_encode(
+        &self,
+        mut pieces: &mut [u8],
+        mut piece_indexes: &[u64],
+    ) -> Result<(), BatchEncodeError> {
+        if pieces.len() % PIECE_SIZE != 0 {
+            return Err(BatchEncodeError::NotMultipleOfPieceSize);
+        }
+
+        #[cfg(feature = "cuda")]
+        if self.cuda_available {
+            let pieces_to_process = pieces.len() / PIECE_SIZE / GPU_PIECE_BLOCK * GPU_PIECE_BLOCK;
```

Comment:
why do we divide then multiply again with `GPU_PIECE_BLOCK`?

Reply:
`2000 / 1024 * 1024 == 1024`